readme.md
# Tk Form

__Project:__ [uom/tk-form](http://packagist.org/packages/uom/tk-form)
__Web:__ <http://www.tropotek.com/>  
__Authors:__ Michael Mifsud <http://www.tropotek.com/>  
  
This is the lib for the Tk Framework HTML Forms

## Contents

- [Installation](#installation)
- [Introduction](#introduction)


## Installation

Available on Packagist ([uom/tk-form](http://packagist.org/packages/uom/tk-form))
and as such installable via [Composer](http://getcomposer.org/).

```bash
composer require uom/tk-form
```

Or add the following to your composer.json file:

```json
"uom/tk-form": "~3.0"
```

If you do not use Composer, you can grab the code from GitHub, and use any
PSR-0 compatible autoloader to load the classes.

## Introduction










changelog.md
#CHANGELOG#

Ver 3.0.112 [2022-02-10]:
-------------------------------


Ver 3.0.110 [2021-12-21]:
-------------------------------


Ver 3.0.108 [2021-10-21]:
-------------------------------


Ver 3.0.106 [2021-09-01]:
-------------------------------


Ver 3.0.104 [2021-05-10]:
-------------------------------


Ver 3.0.102 [2021-04-23]:
-------------------------------


Ver 3.0.100 [2021-04-15]:
-------------------------------


Ver 3.0.98 [2021-03-12]:
-------------------------------


Ver 3.0.96 [2021-03-02]:
-------------------------------
  - Merge branch '3.0' of ssh://github.com/tropotek/tk-form into 3.0


Ver 3.0.94 [2021-03-01]:
-------------------------------


Ver 3.0.92 [2021-02-26]:
-------------------------------
  - Merge branch '3.0' of ssh://github.com/tropotek/tk-form into 3.0


Ver 3.0.90 [2021-02-24]:
-------------------------------


Ver 3.0.88 [2021-02-18]:
-------------------------------
  - NAILED IT!!! New DomRenderer object for the Tk\Form way cleaner code
  - Merge branch '3.0' of ssh://github.com/tropotek/tk-form into 3.0


Ver 3.0.86 [2021-02-12]:
-------------------------------


Ver 3.0.84 [2021-01-19]:
-------------------------------


Ver 3.0.82 [2020-11-23]:
-------------------------------
  - Tagging and releasing branch `3.0` with version `3.0.78`.
  - Tagging and releasing branch `3.0` with version `3.0.76`.
  - Tagging and releasing branch `3.0` with version `3.0.74`.
  - Tagging and releasing branch `3.0` with version `3.0.72`.
  - Tagging and releasing branch `3.0` with version `3.0.70`.
  - Merge branch '3.0' of ssh://github.com/tropotek/tk-form into 3.0
  - Tagging and releasing branch `3.0` with version `3.0.68`.
  - Tagging and releasing branch `3.0` with version `3.0.66`.
  - Removed tkAge jQuery plugin as it interfears with dualListBox plugin
  - Tagging and releasing branch `3.0` with version `3.0.64`.
  - Tagging and releasing branch `3.0` with version `3.0.62`.
  - Tagging and releasing branch `3.0` with version `3.0.60`.
  - Added a form dialog field to TK lib
  - Added new age addon to date input
  - updated PathCase field locations
  - Added colour field to PathCase
  - Updated Weight Field to Path CAse
  - Tagging and releasing branch `3.0` with version `3.0.58`.
  - Tagging and releasing branch `3.0` with version `3.0.56`.
  - Added new placement table action to change status of bulk placements
  - Tagging and releasing branch `3.0` with version `3.0.54`.
  - Tagging and releasing branch `M	Tk/Form/Field/CheckboxGroup.php` with version `3.0.52`.


Ver 3.0.78 [2020-11-10]:
-------------------------------


Ver 3.0.76 [2020-11-09]:
-------------------------------


Ver 3.0.74 [2020-11-03]:
-------------------------------


Ver 3.0.72 [2020-10-26]:
-------------------------------


Ver 3.0.70 [2020-10-22]:
-------------------------------
  - Merge branch '3.0' of ssh://github.com/tropotek/tk-form into 3.0


Ver 3.0.68 [2020-10-22]:
-------------------------------


Ver 3.0.66 [2020-10-21]:
-------------------------------
  - Removed tkAge jQuery plugin as it interfears with dualListBox plugin


Ver 3.0.64 [2020-10-21]:
-------------------------------


Ver 3.0.62 [2020-10-20]:
-------------------------------


Ver 3.0.60 [2020-10-19]:
-------------------------------
  - Added a form dialog field to TK lib
  - Added new age addon to date input
  - updated PathCase field locations
  - Added colour field to PathCase
  - Updated Weight Field to Path CAse


Ver 3.0.58 [2020-05-18]:
-------------------------------


Ver 3.0.56 [2020-05-08]:
-------------------------------
  - Added new placement table action to change status of bulk placements


Ver 3.0.54 [2020-03-09]:
-------------------------------


Ver 3.0.52 [2020-02-25]:
-------------------------------


Ver 3.0.50 [2020-02-14]:
-------------------------------


Ver 3.0.48 [2020-02-11]:
-------------------------------


Ver 3.0.46 [2020-02-04]:
-------------------------------


Ver 3.0.44 [2020-01-29]:
-------------------------------


Ver 3.0.42 [2020-01-29]:
-------------------------------


Ver 3.0.40 [2020-01-28]:
-------------------------------


Ver 3.0.38 [2020-01-23]:
-------------------------------


Ver 3.0.36 [2020-01-23]:
-------------------------------


Ver 3.0.34 [2020-01-23]:
-------------------------------


Ver 3.0.32 [2020-01-21]:
-------------------------------


Ver 3.0.30 [2020-01-13]:
-------------------------------


Ver 3.0.28 [2020-01-07]:
-------------------------------
  - Fixing minor bugs ready for new release
  - Added new course record to the tk-unk lib for ver 3.2


Ver 3.0.26 [2019-12-06]:
-------------------------------


Ver 3.0.24 [2019-12-05]:
-------------------------------


Ver 3.0.22 [2019-12-04]:
-------------------------------


Ver 3.0.20 [2019-11-25]:
-------------------------------
  - Updated status emails and notices for the new Assessment plugin


Ver 3.0.18 [2019-11-18]:
-------------------------------


Ver 3.0.16 [2019-11-12]:
-------------------------------
  - Added more to the new CA plugin


Ver 3.0.14 [2019-09-10]:
-------------------------------
  - Added Min/Max unit value overrides to subjects


Ver 3.0.12 [2019-08-21]:
-------------------------------
  - UPdated the system to use the new dialog box suite


Ver 3.0.10 [2019-07-31]:
-------------------------------


Ver 3.0.8 [2019-06-17]:
-------------------------------
  - In the middle of updating the EMS Institutions pages
  - Fixed multi File input field
  - Cleaning up a lot of managers and editors
  - Fixed Placementreport File upload Issue


Ver 3.0.6 [2019-05-15]:
-------------------------------


Ver 3.0.4 [2019-05-15]:
-------------------------------


Ver 3.0.2 [2019-05-15]:
-------------------------------


Ver 3.0.0 [2019-05-12]:
-------------------------------
  - YEH BABY Symfony base added.


Ver 2.1.76 [2019-04-18]:
-------------------------------


Ver 2.1.74 [2019-03-13]:
-------------------------------


Ver 2.1.72 [2019-02-22]:

js/jquery.tkAddress.js
/*
 * Plugin: Example
 * Version: 1.0
 * Date: 11/05/17
 *
 * @author Michael Mifsud <info@tropotek.com>
 * @see http://www.tropotek.com/
 * @license Copyright 2007 Michael Mifsud
 * @source http://stefangabos.ro/jquery/jquery-plugin-boilerplate-revisited/
 */

/**
 * Use this scrip to manage the company address fields in a form
 *
 * Requires:
 * <code>
 *   $template->appendJsUrl(\Tk\Uri::create('https://maps.googleapis.com/maps/api/js?key='.$gKey.'&libraries=places'));
 *   $template->appendJsUrl(\Tk\Uri::create('/src/App/Form/Field/jquery.gmapSelect.js'));
 *   $template->appendJsUrl(\Tk\Uri::create('/src/App/Controller/Company/jquery.tkAddress.js'));
 * </code>
 *
 * <code>
 *   $(document).ready(function() {
 *     // attach the plugin to an element
 *     $('#element').tkAddress({'foo': 'bar'});
 *
 *     // call a public method
 *     $('#element').data('tkAddress').foo_public_method();
 *
 *     // get the value of a property
 *     $('#element').data('tkAddress').settings.foo;
 *
 *   });
 * </code>
 *
 * @TODO:
 * @TODO:  We should give the user the ability to enter the address manually by a toggle btn or something similar
 * @TODO:
 * @TODO:
 *
 */
(function ($) {
  /**
   *
   * @param element Form
   * @param options Object
   */
  var tkAddress = function (element, options) {
    // plugin vars
    var defaults = {
      gmapSelector: '.tk-gmap-select',
      toggleBtnSelector: '.btn-toggle',
      defaultMode: 'auto',
      manualBtn: false,
      markerIcon: null,
      fieldNames: {
        address: 'address',
        street: 'street',
        city: 'city',
        state: 'state',
        postcode: 'postcode',
        country: 'country',
        lat: 'mapLat',
        lng: 'mapLng',
        zoom: 'mapZoom'
      },
      lang: {
        updateAddress: 'Do you wish to update the address fields to this new location?'
      },

      onUpdateMode: function (plugin) {
        if ($(this).data('editMode') === 'auto') {
          $.each(fields, function () {

            if (
              $(this).is('[name=address]') ||
              $(this).is('[name=mapLat]') ||
              $(this).is('[name=mapLng]') ||
              $(this).is('[name=mapZoom]')
            ) return;
            var parent = $(this).closest('.form-group');
            parent.hide();
          });
          form.find('[name='+plugin.settings.fieldNames[0]+']').show();
        } else {
          $.each(fields, function () {
            if (
              $(this).is('[name=address]') ||
              $(this).is('[name=mapLat]') ||
              $(this).is('[name=mapLng]') ||
              $(this).is('[name=mapZoom]')
            ) return;
            var parent = $(this).closest('.form-group');
            parent.show();
          });
          //form.find('[name='+plugin.settings.fieldNames[0]+']').show();
        }

      }
    };
    var $element = $(element);
    var plugin = this;
    plugin.settings = {};
    var fields = {};
    var autocomplete;
    var form = null;

    // constructor method
    plugin.init = function () {
      plugin.settings = $.extend({}, defaults, $(element).data(), options);
      form = $(element).closest('form');


      // Get all field elements
      for (var fieldName in plugin.settings.fieldNames) {
        var name = plugin.settings.fieldNames[fieldName];
        fields[fieldName] = form.find('[name='+name+']');
        if (!fields[fieldName].length) {
          console.warn('The address auto-complete functions will not work unless all fields are available.');
        }
      }

      if (plugin.settings.manualBtn) {
        $element.data('editMode', plugin.settings.defaultMode);
        // Bind the manual edit toggle button
        $element.parent().find(plugin.settings.toggleBtnSelector).on('click', function () {
          var input = $(this).closest('.form-group').find('input.form-control');
          if (input.data('editMode') === 'auto') {
            input.data('editMode', 'manual');
          } else {
            input.data('editMode', 'auto');
          }
          updateFieldMode();
        });

        // hide address fields by default excecpt the "address" field itself
        // 'auto' use google and hide fields, 'manual' use all address fields
        updateFieldMode();
      } else {
        $element.parent().find(plugin.settings.toggleBtnSelector).parent().hide();
        $element.parent().removeClass('input-group input-append').addClass('form-group');
      }

      if ($element.attr('data-marker-icon') !== undefined && plugin.settings.markerIcon === null) {
        plugin.settings.markerIcon = $element.attr('data-marker-icon');
      }

      // Not the best solution but this stops the form from submitting when enter is hit on this field
      fields.address.on('focus', geolocate).on('keydown', function (e) {
        var code = (e.keyCode ? e.keyCode : e.which);
        if (13 === code) return false;
      });

      // Setup the autocomplete `address` field
      autocomplete = new google.maps.places.Autocomplete(
        (fields.address.get(0)),   /** @type {!HTMLInputElement} */
        {types: ['geocode']}
      );

      // When the user selects an address from the dropdown, populate the address fields in the form.
      autocomplete.addListener('place_changed', function () {
        var place = autocomplete.getPlace();
        updateAddressFields(place);
        if (place.geometry) {
          fields.lat.val(place.geometry.location.lat());
          fields.lng.val(place.geometry.location.lng());
          if (gmapSelect != undefined) {
            gmapSelect.data('gmapSelect').setLatLng(place.geometry.location.lat(), place.geometry.location.lng());
            gmapSelect.data('gmapSelect').redraw();
          }
        }
      });


      // ---------------------------------------
      // Setup gmap select fields javascript
      // ---------------------------------------
      form.find(plugin.settings.gmapSelector + ' .latlng').hide();    // Hide map input fields
      // TODO: Removed this as I am not sure is is still required????
      $.each(fields, function (i, el){
        if (el.attr('name') === plugin.settings.fieldNames['address'] ||
          el.attr('name') === plugin.settings.fieldNames['lat'] ||
          el.attr('name') === plugin.settings.fieldNames['lng'] ||
          el.attr('name') === plugin.settings.fieldNames['zoom']
        ) return;
        // Remove the address fields
        // TODO: See if this is required and if so work out a way to remove those fields.
        //el.parents('.form-row').hide();   // Can remove the submit buttons (See student Change Request in EMS)
        //el.closest('.form-row').hide();   // Can remove the submit buttons (See student Change Request in EMS)
      });

      var canvas = form.find(plugin.settings.gmapSelector + ' .tk-gmap-canvas');
      if (canvas.gmapSelect != undefined) {
        var gmapSelect = canvas.gmapSelect({
          lat: parseFloat(fields.lat.val()),
          lng: parseFloat(fields.lng.val()),
          zoom: parseFloat(fields.zoom.val()),
          icon: plugin.settings.markerIcon,
          onChange: function (lat, lng, zoom) {
            fields.zoom.val(zoom);
          },
          onSelect: function (lat, lng, zoom) {
            if (fields['address'].val() === '' || confirm(plugin.settings.lang.updateAddress)) {
              fields.lat.val(lat);
              fields.lng.val(lng);
              fields.zoom.val(zoom);

              var geo = new google.maps.Geocoder();
              geo.geocode({'location': {lat: lat, lng: lng}}, function (results, status) {
                if (status === 'OK' && results[0]) updateAddressFields(results[0]);
              });
            } else {
              return false;
            }
          }
        });
      }

    };  // END init()

    // Toggle manual/auto mode
    function updateFieldMode() {
      plugin.settings.onUpdateMode.apply(element, [plugin]);
    }


    // update the address fields with the place data
    function updateAddressFields(place) {
      for (var field in fields) {
        if (field === 'zoom' || field === 'lat' || field === 'lng') continue;
        //if (field === 'zoom') continue;
        fields[field].val('');
        fields[field].removeAttr('disabled');
      }
      // Get each component of the address from the place details
      // and fill the corresponding field on the form.
      if (place.address_components) {
        var num = '';
        fields.address.val(place.formatted_address);
        for (var i = 0; i < place.address_components.length; i++) {
          var addressType = place.address_components[i].types[0];
          var value = place.address_components[i]['long_name'];

          if (addressType === 'street_number') {  // Street number
            num = value;
          } else if (addressType === 'route') {    // Street name
            fields.street.val(num + ' ' + value);
          } else if (addressType === 'locality') {   // City
            fields.city.val(value);
          } else if (addressType === 'administrative_area_level_1') {   // state
            fields.state.val(value);
          } else if (addressType === 'country') {   // country
            fields.country.val(value);
          } else if (addressType === 'postal_code') {   // postcode
            fields.postcode.val(value);
          }
        }
      } else if (place.name) {
        fields.address.val(place.name);
      }
      // if (place.geometry) {
      //   fields.lat.val(place.geometry.location.lat());
      //   fields.lng.val(place.geometry.location.lng());
      // }
    }



    // Bias the autocomplete object to the user's geographical location,
    // as supplied by the browser's 'navigator.geolocation' object.
    function geolocate() {
      if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(function (position) {
          var geolocation = {
            lat: position.coords.latitude,
            lng: position.coords.longitude
          };
          var circle = new google.maps.Circle({
            center: geolocation,
            //radius: position.coords.accuracy
            radius: 2000
          });
          autocomplete.setBounds(circle.getBounds());
        });
      }
    }


    // private methods
    //var foo_private_method = function() { };

    // public methods
    //plugin.foo_public_method = function() { };

    // call the "constructor" method
    plugin.init();
  };

  // add the plugin to the jQuery.fn object
  $.fn.tkAddress = function (options) {
    return this.each(function () {
      if (undefined === $(this).data('tkAddress')) {
        var plugin = new tkAddress(this, options);
        $(this).data('tkAddress', plugin);
      }
    });
  }

})(jQuery);
